src/cargo/sources/registry/remote.rs

```diff
@@ -187,84 +186,146 @@ impl<'cfg> RegistryData for RemoteRegistry<'cfg> {
-        handle.get(true)?;
-        handle.url(&url)?;
-        handle.follow_location(true)?;
-        let mut state = Sha256::new();
-        let mut body = Vec::new();
-        network::with_retry(self.config, || {
-            state = Sha256::new();
-            body = Vec::new();
-            let mut pb = Progress::new("Fetch", self.config);
-            {
-                handle.progress(true)?;
-                let mut handle = handle.transfer();
-                handle.progress_function(|dl_total, dl_cur, _, _| {
-                    pb.tick(dl_cur as usize, dl_total as usize).is_ok()
-                })?;
-                handle.write_function(|buf| {
-                    state.update(buf);
-                    body.extend_from_slice(buf);
-                    Ok(buf.len())
-                })?;
-                handle.perform()?;
-            }
-            let code = handle.response_code()?;
-            if code != 200 && code != 0 {
-                let url = handle.effective_url()?.unwrap_or(&url);
-                Err(HttpNot200 { code, url: url.to_string() }.into())
-            } else {
-                Ok(())
-            }
-        })?;
-
-        // Verify what we just downloaded
-        if hex::encode(state.finish()) != checksum {
-            bail!("failed to verify the checksum of `{}`", pkg)
+        let (multi, easy) = self.config.http_multi(to_download.len(), MultiHandler::new)?;
+        let mut running_handles: Vec<_> = (0..easy.len()).map(|_| None).collect();
+        let mut cur_index = 0;
+        // Take the next file in queue and push it into the Multi queue with
+        // the provided Easy. Returns Ok(true) if the handle is configured,
+        // Ok(false) if there's no more files, Err(e) if some error occurred
+        // during configuration. Assumes that the states are clean.
+        let mut configure_handle = |mut handle: Easy2<MultiHandler>, index: usize,
+                                    running_handles: &mut Vec<Option<(usize, Easy2Handle<MultiHandler>)>>| -> CargoResult<bool>{
+            if cur_index == to_download.len() { return Ok(false); }
+            let (pkg_idx, ref url) = to_download[cur_index];
+            let (pkg, _) = pkgs_with_checksum[pkg_idx];
+            cur_index += 1;
+            self.config.shell().status("Downloading", pkg)?;
+            handle.get(true)?;
+            handle.url(&url)?;
+            handle.follow_location(true)?;
+            let mut running_handle = multi.add2(handle)?;
+            running_handle.set_token(index)?;
+            running_handles[index] = Some((pkg_idx, running_handle));
+            Ok(true)
+        };
+        for (index, handle) in easy.into_iter().enumerate() {
+            assert!(configure_handle(handle, index, &mut running_handles)?);
         }
+        loop {
+            // If critical error occurred in multi interface, directly return.
+            let mut active = multi.perform()?;
```

Comment:
I think this method doesn't block in libcurl, so cargo will have to otherwise manage blocking, right?

Reply:
Looks like you're right, thanks for catching it. I will bring in `tokio-curl` to do that.

Reply:
`tokio-curl` has two issues:
- We can't configure the `Multi`. We can probably add a constructor that accepts a `Multi`.
- `Easy2` is not supported. As we have to satisfy the 'static bound in tokio, the allocation is probably inevitable. We probably have to type erase and take `Easy2<Box<dyn Handler>>`.

Reply:
Oh yeah we can definitely add a constructor accepting `Multi`, I think I hadn't gotten around to it yet!

I also haven't currently thought too hard about `Easy2` vs `Easy` wrt tokio-curl, but I'd be more than willing to add support for it